src/openassetio-core/specification/Specification.cpp

```diff
@@ -76,6 +78,8 @@ void Specification::setTraitProperty(const trait::TraitId& traitId,
                                      trait::property::Value propertyValue) {
   impl_->setTraitProperty(traitId, propertyKey, std::move(propertyValue));
 }
+
+bool Specification::operator==(const Specification& other) const { return *impl_ == *other.impl_; }
```

Comment:
[note] I did wonder if we should check memory address equality as a short-circuit optimisation.

It occurred to me that `unordered_map` probably does an identity check to short-circuit its equality operator. So I stepped through the code and, at least in gcc, it doesn't.

Maybe they thought it's such a rare case as to not be worth the cost of adding a pointer comparison to every equality check.  

I suspect that argument applies to us too.

Reply:
I was wondering the very same and was going to chat to you about that! Like your reasoning. I'm sure if it was valuable, then it would be in `unordered_map` too.